src/store/reader.rs

```diff
@@ -93,6 +93,70 @@ impl StoreReader {
     }
 }
 
+pub struct StoreReaderIterator<'a> {
+    block_offset: usize,
+    document_offset: usize,
+    store_reader: &'a StoreReader,
+    uncompressed_block: Vec<u8>,
+}
+
+impl<'a> StoreReaderIterator<'a> {
+    pub fn has_more_documents(&self) -> bool {
+        self.document_offset < self.uncompressed_block.len()
+    }
+
+    pub fn has_more_blocks(&self) -> bool {
+        self.block_offset < self.store_reader.block_data().len()
+    }
+
+    fn read_block(&mut self) -> crate::Result<()> {
+        let (block_size, compressed_block) = self.store_reader.compressed_block(self.block_offset);
+
+        self.uncompressed_block.clear();
+        decompress(&compressed_block, &mut self.uncompressed_block)?;
+
+        self.block_offset += block_size;
+        self.document_offset = 0;
+        Ok(())
+    }
+}
+
+impl<'a> IntoIterator for &'a StoreReader {
+    type Item = Document;
```

Comment:
I am working on a different PR to be more friendly with `io::Read` based Directory. Panicking on data corruption is a choice tantivy makes, but panicking on an `io::Error` is not something we can do (sorry you could not know that).

For this reason, `Iterator<Item=Document>` is not the right API for this reason. 
`Iterator<Item=crate::Result<Document>>` could be ok. (something similar to the `LineIterator` in the standard library)